service/ipam/spec.go

```diff
@@ -0,0 +1,37 @@
+package ipam
+
+import (
+	"context"
+	"net"
+)
+
+// AllocationCallbacks holds function pointers for two methods that must be
+// called inside a lock when allocating new network range.
+type AllocationCallbacks struct {
+	// GetReservedNetworks implementation must return all networks that are
+	// allocated on any given moment. Failing to do that will result in
+	// overlapping allocations.
+	GetReservedNetworks func() ([]net.IPNet, error)
+
+	// PersistAllocatedNetwork must mutate shared persistent state so that on
+	// successful execution persistet network is visible when
+	// GetReservedNetworks() is called.
+	PersistAllocatedNetwork func(net.IPNet) error
+}
+
+// NetworkAllocator is an interface for IPAM implementation that manages
+// network range and subnets for an installation.
+type NetworkAllocator interface {
+	// Allocate performs the subnet allocation from given fullRange for a CIDR
+	// block with netSize. It requires function pointers to callbacks that are
+	// used for getting currently reserved networks and persisting allocated
+	// network. These are called within an single lock so given that callbacks
+	// work correctly, this is concurrent safe within single process e.g. when
+	// logic between multiple controllers are allocating networks from the same
+	// pool.
+	//
+	// NOTE: This is NOT concurrent safe between distinct processes.
+	//
+	// This returns either allocated network or an error.
+	Allocate(ctx context.Context, fullRange net.IPNet, netSize net.IPMask, callbacks AllocationCallbacks) (net.IPNet, error)
```

Comment:
I do not follow that callback thing. Why do we need that? To me it looks like a way to configure IPAM to break. What is the idea behind that?

Reply:
For different clusters there are different AWS clients which affect the way how existing networks are gathered. I don't want to mix handling of those inside the `NetworkAllocator` - it's not really its business.

Similarly persisting allocated network means different thing for old code and new code (whether it's written to `AWSConfig` or `Cluster` CR). This also doesn't belong to `NetworkAllocator`.

In order to have correctness, both gathering existing networks and persisting allocated new subnet must be performed inside a single lock. That's why those implementations are passed as callbacks to this service.

Reply:
With you.